### PR TITLE
Feat 10 & 11

### PR DESCRIPTION
**Pull request by reggieross**
# Description

Implements two new windowing operators—**tumbling** and **sliding**—as part of the `windower` package. Both operators are built on the shared `partition.Base` and integrate with the `pipes.Pipes[R]` abstraction, allowing batch-based aggregation over processing streams.

# Changes summary
- **[FEAT]** Add **KeyedDataStream**
- **[FEAT]** Add **tumbling window** operator (`NewTumbling` + factory)  
- **[FEAT]** Add **sliding window** operator (`NewSliding` + factory)  
- **[CHORE]** Refactor `partition.Base` to use `pipes.Pipes[R]` for multi-channel output  
- **[FIX]** Sliding window: efficient eviction (amortized O(1)) + memory-safe slice reallocation  
- **[TEST]** Comprehensive unit tests for both window types  
- Implements features from:
  - [#11](https://github.com/elastiflow/pipelines/issues/11) – Implement Windowing (Tumbling and Sliding Windows) for KeyedDataStream
  - [#10](https://github.com/elastiflow/pipelines/issues/10) – Implement KeyBy, KeyedDataStream, and Watermarking

### Window Throughput (ns/op)

| Count | ns/op  |
|-------|--------|
| 1     | 33.08  |
| 10    | 31.92  |
| 100   | 30.99  |
| 1000  | 29.95  |

### Varying Buffer Size (ns/op)

| Buffer | ns/op  |
|--------|--------|
| 1      | 532.0  |
| 10     | 424.8  |
| 100    | 309.3  |
| 1000   | 294.4  |

### Key Cardinality (ns/op)

| Keys  | ns/op  |
|-------|--------|
| 1     | 36.41  |
| 2     | 33.09  |
| 10    | 39.60  |
| 100   | 31.40  |
| 1000  | 29.55  |

### Concurrency (ns/op)

| Producers | Partitions | ns/op  |
|-----------|------------|--------|
| 1         | 1          | 31.93  |
| 1         | 4          | 30.09  |
| 1         | 8          | 30.31  |
| 4         | 1          | 29.96  |
| 4         | 4          | 33.55  |
| 4         | 8          | 29.95  |
| 8         | 1          | 30.42  |
| 8         | 4          | 25.55  |
| 8         | 8          | 24.92  |

# QA/testing plan

- [x] Unit tests for `tumbling` and `sliding` windows (including edge cases and error paths)  
- [x] Automated benchmarks in `windower/benchmark_test.go`  
- [ ] Manual smoke test in example pipeline to verify end-to-end aggregation  
- [ ] Skip QA: no user-facing UI changes  

# Reviewer's Checklist

- [ ] Ran linter (`make lint`)  
- [ ] Ran security scan (`make scan:sca`)  
- [ ] All automated tests passed (`make test`)  
- [ ] Updated godoc (`make docs`)  
- [ ] Updated examples where windowing is showcased (`examples/`)

### CODE REVIEW

**Comment by jecsand838** [comment]
@reggieross Nice work getting all of this done! Before I do an in-depth review, could you get the go checks working again?

File: example_window_test.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+			return res, nil
+		}, 500*time.Millisecond) // process over 500ms
+
+		keyedOut := datastreams.Window[int, int, int](
+			datastreams.KeyBy[int, int](
+				p,
+				func(i int) int {
+					return i % 3 // Key by modulo 3
+				},
+			),
```
Comment (jecsand838) [inline example_window_test.go:42]: @reggieross I like what you're doing here. 

Would you mind also including:

1. A example which partitions on a struct type as well 
2. A dedicated `example_partition_test.go` file which dives deeper into partitioning, watermarking, etc.?

Most real world use-cases for partitioning a datastream would involve a using one or more fields on a struct.

File: example_window_test.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		partitionFactory := windower.NewIntervalFactory(func(t []int) (int, error) {
+			if len(t) == 0 {
+				return 0, nil
+			}
+			res := 0
+			for _, elem := range t {
+				res += elem
+			}
+			return res, nil
+		}, 500*time.Millisecond) // process over 500ms
```
Comment (jecsand838) [inline example_window_test.go:34]: @reggieross Was there a reason we deviated from this usage?

```go
ds := datastreams.FromArray([]Event{{GroupID: "a"}, {GroupID: "a"}})
ks := KeyBy(ds, func(e Event) string {return e.GroupID})

// Time base tumbling window
tumbleAgg := Window(ks,
    TumblingWindowOf(5 * time.Second),
    func(batch [Event]) (float64, error) {
        var sum float64
        for _, e := range batch {
            sum += e.Value
        }
        return sum, nil
    },
)

// Count based sliding window
slidingAgg := Window(ks,
    SlidingWindowOfCount(100, 50),
    func(batch [Event]) (int, error) {
        return len(batch), nil
    },
)
```

I'm concerned the factory approach introduces added complexity, couples the windowing logic tighter with the partitioning logic than is needed, and won't extend to later support a `windowAll()` type function on non-partitioned DataStreams and non-time based windowing.

One of the design principles I'd advocate for is keeping our implementation as similar to Flink as possible to make it easy for developers familiar with existing stream processing engines to jump in and get going.

i.e. https://romanglushach.medium.com/apache-flink-windowing-a-deep-dive-into-the-core-feature-of-stream-processing-5fb9198b9a2a

Comment (reggieross) [reply example_window_test.go:34]: @jecsand838 – I think this is a bit of an overstatement.

This was the initial interface you proposed:

```
func Window[T any, K comparable, R any](
    ds KeyedDataStream[T, K],
    spec WindowSpec,
    windowFunc WindowFunc[T, R],
    params ...Params,
) KeyedDataStream[R, K] {
    // ...
}
```

When we started designing the windowing layer, there were a few things I wanted to explicitly handle:
- Partitioning: Routing records to the correct partition, and dynamically creating new partitions per key.
- Channel management: Making sure each partition owns its lifecycle (opening, closing, draining) safely.
- Window lifecycle: Resetting and emitting different types of windows (tumbling, sliding, session) consistently.

That led me to realize: even if a stream isn't keyed (i.e., a global window), it's still technically partitioned — just into a single logical partition.

Because of that, I centralized the partitioning logic into a Partitioner abstraction, instead of tying it tightly to WindowFunc or WindowSpec.

I'll post a followup showing how this pattern still cleanly supports:
- WindowAll() for non-keyed streams
- non-time-based windows like count sliding

We can have a more in depth conversation once I post the updates though

Comment (jecsand838) [reply example_window_test.go:34]: @reggieross 

> When we started designing the windowing layer, there were a few things I wanted to explicitly handle:
> 
> Partitioning: Routing records to the correct partition, and dynamically creating new partitions per key.
> Channel management: Making sure each partition owns its lifecycle (opening, closing, draining) safely.
> Window lifecycle: Resetting and emitting different types of windows (tumbling, sliding, session) consistently.
> That led me to realize: even if a stream isn't keyed (i.e., a global window), it's still technically partitioned — just into a single logical partition.
> 
> Because of that, I centralized the partitioning logic into a Partitioner abstraction, instead of tying it tightly to WindowFunc or WindowSpec.

This is absolutely the correct way to handle it. I'm just wondering why you went with a factory based approach.

Also at a cursory glance the `Window` signature below doesn't seem as decoupled as it should be. 

We shouldn't be required to pass a `keyFunc` into the windowing function for example? Developing a dedicated `windowFunc` would simplify your design greatly imo.

```go
func Window[T any, K comparable, R any](
	keyedDs KeyedDataStream[T, K],
	wf partition.Factory[T, R],
	keyFunc KeyFunc[R, K],
	param ...Params,
) KeyedDataStream[R, K] {
	p := applyParams(param...)

	nextPipe, outChannels := next[R](standard, p, len(keyedDs.inStreams), keyedDs.ctx, keyedDs.errStream, keyedDs.wg)
	kdOut := KeyedDataStream[R, K]{
		DataStream: nextPipe,
		keyFunc:    keyFunc,
		timeMarker: keyedDs.timeMarker,
	}

	pm := partition.NewPartitioner[T, K, R](
		keyedDs.ctx,
		wf,
		keyedDs.errStream,
		keyedDs.timeMarker,
		keyedDs.waterMarker,
	)
```

File: datastreams/internal/pipes/types.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	return s
+}
+
+func (c Pipes[T]) Receivers() Receivers[T] {
+	s := make(Receivers[T], len(c))
+	for i := range c {
+		s[i] = c[i]
+	}
+	return s
+}
```
Comment (jecsand838) [inline datastreams/internal/pipes/types.go:37]: Why nest this into it's own package? jc

File: datastreams/windower/tumbling.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+// NewTumblingFactory constructs a factory function that can be used
+// to create new instances of tumbling. This is useful for
+// creating multiple instances with the same processing function and
+// window duration.
+func NewTumblingFactory[T any](
+	windowDuration time.Duration,
+) partition.Factory[T] {
+	return func(
```
Comment (jecsand838) [inline datastreams/windower/tumbling.go:97]: You could probably just do the `windowDuration` check here

```go
    if windowDuration <= 0 {
        return nil, fmt.Errorf("windowDuration must be > 0 (got %s)", windowDuration)
    }
```

File: datastreams/windower/tumbling.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	return t.started
+}
+
+func (t *status) set(state bool) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.started = state
+}
```
Comment (jecsand838) [inline datastreams/windower/tumbling.go:33]: I'd recommend using a `atomic.Bool` here:

```suggestion
import (
	"context"
	"sync"
	"sync/atomic"
	"time"

	"github.com/elastiflow/pipelines/datastreams/internal/partition"
	"github.com/elastiflow/pipelines/datastreams/internal/pipes"
)

type status struct {
	started atomic.Bool
}

func (s *status) running() bool {
	return s.started.Load()
}

func (s *status) SetRunning(v bool) {
	s.started.Store(v)
}
```

File: datastreams/windower/tumbling.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+// tumbling buffers items for windowDuration once the first item arrives.
+// After windowDuration, it publishes the batch and clears it, then waits
+// for the next item to start a new timer.
+type tumbling[T any] struct {
+	*partition.Base[T]
```
Comment (jecsand838) [inline datastreams/windower/tumbling.go:39]: It probably makes more sense to embed this. 

```suggestion
	partition.Base[T]
```

Saves us an allocation and pointer hop while decreasing the GC footprint.

File: datastreams/windower/tumbling.go [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+// Push adds to the current batch; if no timer is running, start one.
+func (t *tumbling[T]) Push(item T) {
+	t.Base.Push(item)
+
+	if !t.status.running() {
+		t.status.set(true)
+		go t.waitAndFlush()
+	}
```
Comment (jecsand838) [inline datastreams/windower/tumbling.go:73]: I'd use `tryStart` to atomically check and set the running status.

```suggestion
	if t.status.tryStart() {
		go t.waitAndFlush()
	}
```

File: datastreams/windower/tumbling.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+// waitAndFlush sleeps windowDuration, then flushes whatever is in the batch.
+func (t *tumbling[T]) waitAndFlush() {
+	timer := time.NewTimer(t.windowDuration)
+	defer timer.Stop()
+
+	select {
+	case <-t.ctx.Done():
+		return
```
Comment (jecsand838) [inline datastreams/windower/tumbling.go:83]:
```suggestion
		// Ensure status is reset if the context is cancelled
		t.status.set(false)
		return
```

File: datastreams/windower/sliding.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+		window[i] = rec.val
+	}
+	if final {
+		s.items = nil
+	}
+
+	s.mu.Unlock()
+	// publish
+	return window
+}
```
Comment (jecsand838) [inline datastreams/windower/sliding.go:76]: I think this code maybe O(n)

```go
	for cut < len(s.items) && s.items[cut].ts.Before(threshold) {
		cut++
	}
```

Consider this tweaked version

```suggestion
func (s *slidingBatch[T]) next(
	windowDuration time.Duration,
	now time.Time,
	final bool,
) []T {
	threshold := now.Add(-windowDuration)
	s.mu.Lock()
	// find cut index via binary search
	cut := sort.Search(len(s.items), func(i int) bool {
		return !s.items[i].ts.Before(threshold)
	})
	if cut > 0 {
		// reallocate so we don't hold old large backing arrays
		tail := s.items[cut:]
		newBuf := make([]record[T], len(tail))
		copy(newBuf, tail)
		s.items = newBuf
	}
	if len(s.items) == 0 {
		s.mu.Unlock()
		return []T{}
	}
	// snapshot values
	window := make([]T, len(s.items))
	for i, rec := range s.items {
		window[i] = rec.val
	}
	if final {
		s.items = nil
	}
	s.mu.Unlock()
	// publish
	return window
}
```

File: datastreams/windower/sliding.go [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+	out pipes.Senders[[]T],
+	errs chan<- error,
+	windowDuration, slideInterval time.Duration,
+) partition.Partition[T] {
+	if windowDuration <= 0 || slideInterval <= 0 {
+		panic("interval and slideInterval must be > 0")
+	}
+	if slideInterval > windowDuration {
+		panic("slideInterval cannot be larger than interval")
+	}
```
Comment (jecsand838) [inline datastreams/windower/sliding.go:100]: I think we should catch and return these errors in `NewSlidingFactory ` instead of panicking.

File: datastreams/windower/interval.go [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+// arrive in a window, we publish an empty batch.
+func NewInterval[T any](
+	ctx context.Context,
+	out pipes.Senders[[]T],
+	errs chan<- error,
+	interval time.Duration,
+) partition.Partition[T] {
+	if interval <= 0 {
+		panic("interval must be > 0")
+	}
```
Comment (jecsand838) [inline datastreams/windower/interval.go:27]: Same with this panic as the others imo.

File: datastreams/windower/interval.go [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+// NewIntervalFactory constructs a factory function that can be used
+// to create new instances of timedInterval. This is useful for
+// creating multiple instances with the same processing function and
+// interval duration.
+func NewIntervalFactory[T any](
+	interval time.Duration,
+) partition.Factory[T] {
+	return func(
```
Comment (jecsand838) [inline datastreams/windower/interval.go:60]: I'd catch the invalid `interval` here and return early imo.

File: datastreams/windower/sliding.go [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+	for {
+		select {
+		case <-ctx.Done():
+			// final, non‑overlapping flush
+			next := s.sb.next(s.windowDuration, time.Now(), true)
+			if len(next) == 0 {
+				continue
+			}
+			s.Flush(ctx, next)
+
```
Comment (jecsand838) [inline datastreams/windower/sliding.go:131]: Is there a risk of goroutine leaking here due to not returning?

File: datastreams/windower/tumbling.go [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+// after a flush will start a new window timer.
+func newTumbling[T any](
+	ctx context.Context,
+	out pipes.Senders[[]T],
+	errs chan<- error,
+	windowDuration time.Duration,
+) partition.Partition[T] {
+	if windowDuration <= 0 {
+		panic("interval must be > 0")
+	}
```
[the remark this answers is not in the document]
Comment (reggieross) [reply datastreams/windower/tumbling.go:55]: We said new constructors shouldn't return an error. I'd either like to panic on have an init func that returns an error

File: datastreams/windower/tumbling.go (hunk 4 above)
Comment (reggieross) [reply datastreams/windower/tumbling.go:97]: same comment as above: 

"We said new constructors shouldn't return an error. I'd either like to panic on have an init func that returns an error"

File: datastreams/windower/sliding.go [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+// sliding emits overlapping windows of size windowDuration,
+// every slideInterval apart.
+type sliding[T any] struct {
+	*partition.Base[T]
+
+	windowDuration time.Duration
+	slideInterval  time.Duration
+	sb             *slidingBatch[T]
+}
```
[the remark this answers is not in the document]
Comment (reggieross) [reply datastreams/windower/sliding.go:86]: What is the buffer pool doing for me here? The flush method just winds the window to the appropriate out channels.

File: datastreams/windower/sliding.go (hunk 9 above)
Comment (reggieross) [reply datastreams/windower/sliding.go:76]: For clarity, the only piece that changed was:

```
cut := sort.Search(len(s.items), func(i int) bool {
  return !s.items[i].ts.Before(threshold)
})
```

right?

File: datastreams/windower/sliding.go (hunk 13 above)
Comment (reggieross) [reply datastreams/windower/sliding.go:131]: yeah I missed this ty

File: datastreams/internal/partition/store.go [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+	w, ok := s.partitions[k]
+	return w, ok
+}
+
+func (s *store[T, K]) set(k K, w Partition[T]) Partition[T] {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.partitions[k] = w
+	return w
+}
```
[the remark this answers is not in the document]
Comment (reggieross) [reply datastreams/internal/partition/store.go:30]: @jecsand838 I can tackle this in a follow up PR since this in an enhancement not a blocker for my implementation

File: datastreams/windower/tumbling.go (hunk 4 above)
Comment (jecsand838) [reply datastreams/windower/tumbling.go:92]: Because this is returning a function and not a struct, I think we really should return an error in this case.

File: datastreams/windower/sliding.go (hunk 9 above)
Comment (jecsand838) [reply datastreams/windower/sliding.go]: Correct

File: datastreams/internal/partition/store.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (jecsand838) [reply datastreams/internal/partition/store.go:30]: @reggieross I made a follow-up ticket to capture this work: [Implement Sharded Store](https://github.com/elastiflow/pipelines/issues/18)

[the remark this answers is not in the document]
Comment (reggieross) [reply datastreams/internal/partition/store.go:30]: tyty

File: datastreams/windower/sliding.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (jecsand838) [reply datastreams/windower/sliding.go:87]: @reggieross Because `next()` is making a new `[]T` each time the ticker fires, the code is allocating a new backing array the size of the current window, copying the window's values into it, and then after the `Flush` the GC has to scan and free that memory. 

A `sync.Pool` lets us re-use the slice which significantly improves performance when done correctly. 

- Here's the results from a basic benchmark I put together to showcase this:

Screenshot 2025-05-14 at 3 23 54 PM

- Benchmark Code

```go
package windower

import (
	"sort"
	"sync"
	"testing"
	"time"
)

var benchSink int

type recInt struct {
	val int
	ts  time.Time
}

// NO slice pooling

type batchNoPool struct {
	items []recInt
}

func newBatchNoPool() *batchNoPool {
	return &batchNoPool{items: make([]recInt, 0, 64)}
}

func (b *batchNoPool) push(v int, ts time.Time) {
	b.items = append(b.items, recInt{v, ts})
}

func (b *batchNoPool) next(windowDur time.Duration, now time.Time) []int {
	threshold := now.Add(-windowDur)
	cut := sort.Search(len(b.items), func(i int) bool {
		return !b.items[i].ts.Before(threshold)
	})
	b.items = b.items[cut:]
	if len(b.items) == 0 {
		return nil
	}
	out := make([]int, len(b.items))
	for i, rec := range b.items {
		out[i] = rec.val
	}
	return out
}

// WITH slice pooling

type batchPool struct {
	items []recInt
	pool  *sync.Pool
}

func newBatchPool(p *sync.Pool) *batchPool {
	return &batchPool{items: make([]recInt, 0, 64), pool: p}
}

func (b *batchPool) push(v int, ts time.Time) {
	b.items = append(b.items, recInt{v, ts})
}

func (b *batchPool) next(windowDur time.Duration, now time.Time) []int {
	threshold := now.Add(-windowDur)
	cut := sort.Search(len(b.items), func(i int) bool {
		return !b.items[i].ts.Before(threshold)
	})
	b.items = b.items[cut:]
	if len(b.items) == 0 {
		return nil
	}
	var out []int
	if v := b.pool.Get(); v != nil {
		out = v.([]int)
	}
	if cap(out) < len(b.items) {
		out = make([]int, len(b.items))
	}
	out = out[:len(b.items)]
	for i, rec := range b.items {
		out[i] = rec.val
	}
	return out
}

func (b *batchPool) put(buf []int) {
	b.pool.Put(buf)
}

//
// Benchmarking
//

func BenchmarkSlidingBuffer(b *testing.B) {
	const (
		windowSize = 10_000            // records visible at once
		slideCount = 1_000              // ticks per benchmark iteration
		windowDur  = time.Second  // 1-second window
		slideDur   = 100 * time.Millisecond
	)
	base := time.Unix(0, 0)
	timestamps := make([]time.Time, windowSize)
	for i := range timestamps {
		timestamps[i] = base.Add(time.Duration(i) * time.Millisecond)
	}
	run := func(withPool bool, parallel bool) func(*testing.B) {
		return func(b *testing.B) {
			var p *sync.Pool
			if withPool {
				p = &sync.Pool{New: func() any { return make([]int, 0) }}
			}
			b.ReportAllocs()
			b.ResetTimer()
			if parallel {
				b.RunParallel(func(pb *testing.PB) {
					for pb.Next() {
						runOneIter(withPool, p, timestamps, windowDur, slideDur, slideCount)
					}
				})
			} else {
				for i := 0; i < b.N; i++ {
					runOneIter(withPool, p, timestamps, windowDur, slideDur, slideCount)
				}
			}
		}
	}
	b.Run("NoPool_Serial", run(false, false))
	b.Run("WithPool_Serial", run(true, false))
	b.Run("NoPool_Parallel", run(false, true))
	b.Run("WithPool_Parallel", run(true, true))
}

func runOneIter(
	withPool bool,
	pool *sync.Pool,
	ts []time.Time,
	windowDur, slideDur time.Duration,
	slideCount int,
) {
	var sb interface {
		push(int, time.Time)
		next(time.Duration, time.Time) []int
	}
	if withPool {
		sb = newBatchPool(pool)
	} else {
		sb = newBatchNoPool()
	}
	for i := 0; i < len(ts); i++ {
		sb.push(i, ts[i])
	}
	now := ts[0].Add(windowDur)
	for i := 0; i < slideCount; i++ {
		buf := sb.next(windowDur, now)
		if withPool && buf != nil {
			pool.Put(buf)
		}
		benchSink += len(buf)
		now = now.Add(slideDur)
	}
}
```

[the remark this answers is not in the document]
Comment (jecsand838) [reply datastreams/windower/sliding.go:87]: @ndesai96 There's also others places internally we could apply this imo.

[the remark this answers is not in the document]
Comment (reggieross) [reply datastreams/windower/sliding.go:87]: ty, that makes sense! Can I do it in in a separate PR then? I can do a performance tuning PR where I update this callout and [18](https://github.com/elastiflow/pipelines/issues/18)

[the remark this answers is not in the document]
Comment (jecsand838) [reply datastreams/windower/sliding.go:87]: @reggieross Sounds good to me, here's the follow-up issue: https://github.com/elastiflow/pipelines/issues/19

File: datastreams/errors.go [hunk 17]
```diff
@@ -24,6 +26,8 @@ func (w ErrorCode) String() string {
 		"MAP",
 		"SINK",
 		"EXPAND",
+		"KEY_BY",
+		"WINDOW",
 	}[w]
 }
```
Comment (jecsand838) [inline datastreams/errors.go:32]: @reggieross Can we can and document the following functions in this file `IsKeyByError()` and `IsWindowError()`? Just to maintain the current pattern and be consistent for our users.

File: datastreams/types.go [hunk 18]
```diff
@@ -15,38 +15,6 @@ type ExpandFunc[T any, U any] func(T) ([]U, error)
 // This function type is used to filter a given input type
 type FilterFunc[T any] func(T) (bool, error)
 
-type receivers[T any] []<-chan T
-type senders[T any] []chan<- T
-type pipes[T any] []chan T
+type KeyFunc[T any, K comparable] func(T) K
```
Comment (jecsand838) [inline datastreams/types.go:18]: @reggieross Can we add a GoDoc comment for `KeyFunc`?

File: datastreams/types.go [hunk 19]
```diff
@@ -15,38 +15,6 @@ type ExpandFunc[T any, U any] func(T) ([]U, error)
...
-}
-
-func (c pipes[T]) Receivers() receivers[T] {
-	s := make(receivers[T], len(c))
-	for i := range c {
-		s[i] = c[i]
-	}
-	return s
-}
+type WindowFunc[T any, R any] func([]T) (R, error)
```
Comment (jecsand838) [inline datastreams/types.go:20]: @reggieross Can we add a GoDoc comment for `WindowFunc`?

**Comment by jecsand838** [comment]
@reggieross LMGTM. 

Only remaining items are:
- Update `README.md` to include information on the new `Window`, `Watermark`, and `KeyedDatastream` features.
- Add the GoDoc comments in `KeyFunc` and `WindowFunc`
- Add those `IsWindowError()` and `IsKeyByError` helper functions.

File: pipeline.go [hunk 20]
```diff
@@ -33,6 +33,8 @@ import (
 // a StreamFunc is another DataStream that can be further chained.
 type StreamFunc[T any, U any] func(stream datastreams.DataStream[T]) datastreams.DataStream[U]
 
+type WindowFunc[T any, U any] func(stream datastreams.DataStream[[]T]) datastreams.DataStream[U]
```
Comment (jecsand838) [inline pipeline.go:36]: @reggieross Can we just remove this one for now? 

I don't see it being used and there's two `WindowFunc` defined.
```suggestion
```

File: pipeline.go [hunk 21]
```diff
@@ -33,6 +33,8 @@ import (
...
+
```
Comment (jecsand838) [inline pipeline.go:37]: @reggieross Can we just remove this one for now?

I don't see it being used and there's two `WindowFunc` defined.

```suggestion
```

**Review by jecsand838 — APPROVED** [review]
